view/frontend/web/internals/common.js

```diff
@@ -1,3 +1,54 @@
+var algolia = {
+	allowedHooks: [
+		'modifyAutocompleteSources',
+		'modifyAutocompleteOptions',
+		'algoliaHookBeforeInstantsearchInit',
+		'algoliaHookBeforeWidgetInitialization',
+		'algoliaHookBeforeInstantsearchStart',
+		'algoliaHookAfterInstantsearchStart'
+	],
+	registeredHooks: [],
+	registerHook: function (hookName, callback) {
+		if (this.allowedHooks.indexOf(hookName) === -1) {
+			throw 'Hook "' + hookName + '" cannot be defined. Please use one of ' + this.allowedHooks.join(', ');
+		}
+		
+		if (this.registeredHooks[hookName] === undefined) {
+			this.registeredHooks[hookName] = [];
+		}
+		
+		this.registeredHooks[hookName].push(callback);
+	},
+	getRegisteredHooks: function(hookName) {
+		if (this.registeredHooks[hookName] === undefined) {
+			return [];
+		}
+		
+		return this.registeredHooks[hookName];
+	},
+	triggerHooks: function () {
+		var hookName = arguments[0];
+		delete arguments[0];
+		
+		var clearedArguments = this.clearArguments(arguments);
+		
+		var registeredHooks = this.getRegisteredHooks(hookName);
+		for (var i = 0; i < registeredHooks.length; i++) {
+			clearedArguments[0] = registeredHooks[i].apply(this, clearedArguments);
+		}
+		
+		return clearedArguments[0];
+	},
+	clearArguments: function(arguments) {
+		var clearArguments = Array.prototype.slice.call(arguments);
```

Comment:
I wouldn't do the transformation from arguments to array here. I'd put it in the calling function and would only use arrays internally.